### PR TITLE
Add the umfPoolCreateEx internal function

### PR DESCRIPTION
**Pull request by kswiecicki**
... that creates a pool owning the provided memory provider.

### CODE REVIEW

**Comment by igchor** [comment]
FYI, this (or something similar) is needed if we want to implement `umfCreateMemoryPoolFromMemspace`.

**Comment by vinser52** [comment]
I understand that this function is needed to implement `umfCreateMemoryPoolFromMemspace ` function, but it should be a public interface?

**Comment by bratpiorka** [comment]
I would prefer a function that combines pool and provider params so the provider would be both created and destroyed with the pool. Something like:

umfPoolCreateEx(const struct umf_memory_pool_ops_t *pool_ops, void *pool_params, 
				const struct umf_memory_provider_ops_t *provider_ops, void *provider_params, 
				umf_memory_pool_handle_t *hPool);
				
with this API we don't expose the handle to the provider to the user at all so it is obvious who is managing and controlling the live of the provider.

**Comment by bratpiorka** [comment]
Yes, I agree that we shouldn't expose this in public API if this is not needed. But also I think that we should follow the RAII concept - if the pool is an owner of the provider, the handle of provider should be hidden so no one could destroy the provider in the background

**Comment by vinser52** [comment]
Will you add the unit test?

**Review by vinser52 — APPROVED** [review]

File: src/memory_pool.c [hunk 1]
```diff
@@ -15,6 +15,30 @@
 #include <assert.h>
 #include <stdlib.h>
 
+enum umf_result_t
```
Comment (bratpiorka) [inline src/memory_pool.c:18]: Note that with recent Łukasz D. changes there is no need to use "enum" and "struct" keywords when you reference umf_result_t etc. Please remove them.

File: src/memory_pool.c [hunk 2]
```diff
@@ -15,6 +15,30 @@
...
+umfPoolCreateEx(const struct umf_memory_pool_ops_t *pool_ops, void *pool_params,
+                const struct umf_memory_provider_ops_t *provider_ops,
+                void *provider_params, umf_memory_pool_handle_t *hPool) {
+    umf_memory_provider_handle_t provider = NULL;
```
Comment (bratpiorka) [inline src/memory_pool.c:22]: Check for pointers != null

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: src/memory_pool.c (hunk 2 above)
Comment (kswiecicki) [reply src/memory_pool.c:27]: I've added a null-check for the output handle pointer and asserts after provider/pool creation.
From what I've seen in `umfPoolCreate` it null-checks only the input provider handle, whereas `umfMemoryProviderCreate` doesn't check any parameters. Should we validate `...provider/pool_ops_t` pointers and input/output handle pointers in those functions?

**Comment by kswiecicki** [comment]
> Will you add the unit test?

I've added test that checks whether `poolCreateEx` creates a valid pool internally and some basic successful pool creation, param fail tests.
@vinser52 @bratpiorka what do you think about adding AddressSanitizer or Valgrind for UMF testing? I could use it to check whether `poolCreateEx` appropriately destroyed the internal memory provider.

**Comment by igchor** [comment]
> > Will you add the unit test?
> 
> I've added test that checks whether `poolCreateEx` creates a valid pool internally and some basic successful pool creation, param fail tests. @vinser52 @bratpiorka what do you think about adding AddressSanitizer or Valgrind for UMF testing? I could use it to check whether `poolCreateEx` appropriately destroyed the internal memory provider.

I think you could try enabling sanitizer, it should be pretty easy (just need to add flag to the compilation).

**Comment by bratpiorka** [comment]
> > > Will you add the unit test?
> > 
> > 
> > I've added test that checks whether `poolCreateEx` creates a valid pool internally and some basic successful pool creation, param fail tests. @vinser52 @bratpiorka what do you think about adding AddressSanitizer or Valgrind for UMF testing? I could use it to check whether `poolCreateEx` appropriately destroyed the internal memory provider.
> 
> I think you could try enabling sanitizer, it should be pretty easy (just need to add flag to the compilation). And you can even reuse logic from UR: https://github.com/oneapi-src/unified-runtime/blob/4424195c7d5d990901f7343fa9e7ac500b03e38d/cmake/helpers.cmake#L45

We had a plans to add various sanitizers - look for them in our JIRA (UMFW-130)

File: src/memory_pool.c (hunk 2 above)
Comment (bratpiorka) [reply src/memory_pool.c:27]: Please validate all input pointers. About other functions - we should add similar checks there, probably in separate PRs. Please note that we'll have to do it anyway when we enable coverity :)

**Review by bratpiorka — CHANGES_REQUESTED** [review]

File: src/memory_pool.c (hunk 2 above)
Comment (kswiecicki) [reply src/memory_pool.c:27]: Done, I've omitted the params, since pool/provider implementations not necessarily need a parameter for its initialization.

**Comment by kswiecicki** [comment]
> > > > Will you add the unit test?
> > > 
> > > 
> > > I've added test that checks whether `poolCreateEx` creates a valid pool internally and some basic successful pool creation, param fail tests. @vinser52 @bratpiorka what do you think about adding AddressSanitizer or Valgrind for UMF testing? I could use it to check whether `poolCreateEx` appropriately destroyed the internal memory provider.
> > 
> > 
> > I think you could try enabling sanitizer, it should be pretty easy (just need to add flag to the compilation). And you can even reuse logic from UR: https://github.com/oneapi-src/unified-runtime/blob/4424195c7d5d990901f7343fa9e7ac500b03e38d/cmake/helpers.cmake#L45
> 
> We had plans to add various sanitizers - look for them in our JIRA (UMFW-130). If you want to take care of these please do it in separate PR.

I've assigned myself for this task.

**Review by bratpiorka — APPROVED** [review]

**Pull request merged**